src/bus.rs

```diff
@@ -158,77 +158,99 @@ impl Bus {
         self.timer.is_interrupt = (value >> TIMER_SHIFT)  & 0x4 != 0;
     }
 
-    fn load(&self, addr: u16) -> Result<u8, ()> {
+    fn find_device(&self, addr: u16) -> Option<&dyn Device> {
+        let ret: &dyn Device;
         match addr {
-            CATRIDGE_START ..= CATRIDGE_END => self.catridge.load(addr),
-            VRAM_START ..= VRAM_END => self.gpu.load(addr),
-            RAM_START ..= RAM_END => self.ram.load(addr),
-            OAM_START ..= OAM_END => self.gpu.load(addr),
-            UNUSABLE_START ..= UNUSABLE_END => {
-                info!("Load at unusable address {:#X}", addr);
-                Ok(0)
-            }
-            HRAM_START ..= HRAM_END => self.hram.load(addr),
-            TIMER_START ..= TIMER_END => self.timer.load(addr),
-            JOYPAD_ADDR => self.joypad.load(addr),
-            INT => Ok(self.load_interrupt()),
-            INTENB => Ok(u8::from(&self.interruptenb)),
-            _ => {
-                // match IO line
-                match FromPrimitive::from_u16(addr) {
-                    Some(IO::LCDC) => Ok(self.gpu.lcdc.to_u8()),
-                    Some(IO::SCY) => Ok(self.gpu.scy),
-                    Some(IO::SCX) => Ok(self.gpu.scx),
-                    Some(IO::LY) => Ok(self.gpu.line),
-                    Some(IO::BGP) => Ok(self.gpu.bg_palette),
-                    Some(IO::OBP0) => Ok(self.gpu.ob0_palette),
-                    Some(IO::OBP1) => Ok(self.gpu.ob1_palette),
-                    Some(_) => {
-                        info!("Unimplemented load on address {:#X}", addr);
-                        Ok(0)
-                    },
-                    None => {
-                        error!("Invalid load on address {:#X}", addr);
-                        Err(())
+            CATRIDGE_START ..= CATRIDGE_END => ret = &self.catridge,
+            VRAM_START ..= VRAM_END => ret = &self.gpu,
+            RAM_START ..= RAM_END => ret = &self.ram,
+            OAM_START ..= OAM_END => ret = &self.gpu,
+            HRAM_START ..= HRAM_END => ret = &self.hram,
+            TIMER_START ..= TIMER_END => ret = &self.timer,
+            JOYPAD_ADDR => ret = &self.joypad,
+            _ => return None,
+        }
+        Some(ret)
+    }
+
+    fn load(&self, addr: u16) -> Result<u8, ()> {
+        match self.find_device(addr) {
+            Some(dev) => dev.load(addr),
+            None => match addr {
+                UNUSABLE_START ..= UNUSABLE_END => {
```

Comment:
After your change of memory permission flag, make UNUSALBE a Memory struct and mark it as invalid.
We want to put any possible region into your find_device.

Reply:
According to my test binary trying to access some UNUSABLE memory, so I marked the area as NORMAL rather than  Invalid.

Reply:
I read your comment clearly once again, and I found out that I was misunderstood about your description of `Invalid memory`. I am going to push another commit to fix it.